### ISSUE TITLE
BYO LLVM build configuration broken by pybind11 changes

### ISSUE THREAD

**Issue by ScottTodd** [iree-org/iree#18884]
See https://github.com/iree-org/iree/pull/18872#issuecomment-2435492981

Logs: https://github.com/iree-org/iree/actions/runs/11496514385/job/31998278068#step:4:10340

```
-- Found Python3: /__w/iree/iree/build-byo-llvm/.venv/bin/python3 (found version "3.11.0") found components: Interpreter Development Development.Module Development.Embed 
-- Found Python3: /__w/iree/iree/build-byo-llvm/.venv/bin/python3 (found version "3.11.0") found components: Interpreter Development.Module 
-- Found python include dirs: /usr/include/python3.11
-- Found python libraries: /usr/lib/x86_64-linux-gnu/libpython3.11.so
-- Found numpy v: 
-- Checking for pybind11 in python path...
-- not found (install via 'pip install pybind11' or set pybind11_DIR)
CMake Error at cmake/modules/MLIRDetectPythonEnv.cmake:28 (find_package):
  Could not find a package configuration file provided by "pybind11"
  (requested version 2.9) with any of the following names:

    pybind11Config.cmake
    pybind11-config.cmake

  Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
  "pybind11_DIR" to a directory containing one of the above files.  If
  "pybind11" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:175 (mlir_configure_python_dev_packages)


-- Configuring incomplete, errors occurred!
```

Code:

* https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_x64_clang_byollvm.yml
* https://github.com/iree-org/iree/blob/main/build_tools/cmake/build_and_test_byo_llvm.sh
* https://github.com/iree-org/iree/blob/main/build_tools/llvm/byo_llvm.sh
* https://github.com/iree-org/iree/blob/main/build_tools/cmake/iree_llvm.cmake

---

We can also increase the frequency of https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_x64_clang_byollvm.yml (from nightly to every commit) if we have CI budget/capacity, and add alerting (like posting to Discord on failure).

**Comment by marbre** [comment]
The problem here is that `build_tools/cmake/build_and_test_byo_llvm.sh` builds MLIR before IREE
https://github.com/iree-org/iree/blob/03c744ead1482abde3ee9e70293215c5b557c629/build_tools/cmake/build_and_test_byo_llvm.sh#L37-L38
whereas the latter fetches pybind and sets `pybind11_DIR` accordingly. One solution is to replicate https://github.com/iree-org/iree/blob/03c744ead1482abde3ee9e70293215c5b557c629/CMakeLists.txt#L778-L796

Other options could be to switch back to a submodule and to
* modify the way MLIR detects pybind _or_
* fix the `find_package` integration in pybind upstream _or_
* migrate to a newer CMake to enable `FIND_PACKAGE_ARGS` / `OVERRIDE_FIND_PACKAGE` (issue for CMake migration #18598) .

**Comment by stellaraccident** [comment]
The submodule was broken. Let's not do that. 

To unblock, just pip install a matching pybind11 locally in the byo script for now. Leave a comment about it and let it be.

**Comment by marbre** [comment]
> The submodule was broken. Let's not do that.
> 
> To unblock, just pip install a matching pybind11 locally in the byo script for now. Leave a comment about it and let it be.

Hm, I though it was the combination of the submodule + a pip installed version rather than the submodule. With a pip install in the BYO script which than later gets the source via FetchContent we combine providing the source alongside to a pip install again but I have no strong feelings as long as this does not breaks us again to soon.

**Comment by stellaraccident** [comment]
Getting precise version sync between a pip install and a submodule is unreliable for normal dev because there is nothing forcing people to upgrade their requirements. 

For something low priority and weird like the byo flow, there is already a lot of bespoke in there. A little bit more isn't the end of the world.
